### ISSUE TITLE
Story Cycles With Conditional Checkpoints Aren't Trained Properly

### ISSUE THREAD

**Issue by jctrouble** [RasaHQ/rasa_core#166]
**Rasa Core version**:
HEAD (57e7a8928a598e6af443767b0f92faa7937da10c)

**Python version**: 
Python 3.6.3

**Operating system** (windows, osx, ...):
OSX, 10.13.2

**Issue**:
**tl;dr;** It seems that stories that utilize conditional checkpoints and have cycles don't seem to properly train, with features/labels not being emitted for entire story steps that are started with a conditional checkpoint.

I applied the following [patch](https://github.com/RasaHQ/rasa_core/files/1621335/logging.patch.txt) to generator.py to log out some internal state when generating the story (the patch isn't particularly interesting, it's really just for reference for the next gist):
https://gist.github.com/jctrouble/e12d81172759cc6a537fb8bde4e04cce

When I go to train my model (history=5), I got the following output:
https://gist.github.com/jctrouble/acba9e901a08bfbd7a8786479687851d

Basically, no features/labels are emitted for the story step starting with `> process_name{"name":"Josh"}`.

Notes:
- In the example given above, I commented out the extra augmentation rounds to help keep things simple, but even when I re-enabled the augmentation, the branches were skipped.
- I have this issue for real in a much more complex story for my own project, this example is really more of a proof of concept. The issue manifested basically as an entire conditional branch of our story being ignored, and at runtime the model would predict the other case of the conditional with near-100% confidence.
- When exploring this problem, I would see variance in between runs of the generator, with the location of `CYCLE_<blah>` checkpoints not quite being deterministic (and leading to a different set of features/labels being generated). I assumed this was due to a dictionary (and it's lack of ordering) being involved in the DAG. Dunno if that is in play here or not, but I thought it interesting to note.

**Content of [domain file](https://github.com/RasaHQ/rasa_core/files/1621341/domain.yaml.txt)**:
```yaml
action_factory: remote
slots:
  name:
    type: text
intents:
  - greet
  - inform_name
entities:
 - name
actions:
  - action_lookup_name
  - action_request_name
  - action_repeat_name
```

**Content of [story file](https://github.com/RasaHQ/rasa_core/files/1621342/stories.md.txt)**:
```
## Greeting
* greet{}
- action_request_name
> get_name

## Get Name - None
> get_name
* inform_name{"name":null}
> process_name

## Get Name - Value
> get_name
* inform_name{"name":"Josh"}
> process_name

## Have name
> process_name{"name":"Josh"}
- action_repeat_name
- action_restart

## Dont have name
> process_name{"name":null}
- action_request_name
> get_name
```

[logging.patch.txt](https://github.com/RasaHQ/rasa_core/files/1621335/logging.patch.txt)
[domain.yaml.txt](https://github.com/RasaHQ/rasa_core/files/1621341/domain.yaml.txt)
[stories.md.txt](https://github.com/RasaHQ/rasa_core/files/1621342/stories.md.txt)

**Comment by tmbo** [comment]
Thanks a lot for this detailed report - I will try to reproduce that and let you know how that goes.

**Comment by znat** [comment]
This makes me think of an issue I had: some stories were ignored because they were put out of the training set and added to the validation set. Have you tried setting the `validation_split` to 0 in your training config?
```
agent.train(
            training_data_file,
            max_history=3,
            epochs=200,
            batch_size=50,
            augmentation_factor=100,
            validation_split=0.0
    )
```

**Comment by jctrouble** [comment]
@znat Thanks for the thoughtful suggestion. The particular example I gave above was occurring during story generation time – training data wasn't even being created (much less getting segmented off for validation). In my real project (rather than this toy example), this problem would manifest even with 100 training epochs (i.e. enough that I would expect to compensate for the validation split).

**Comment by tmbo** [comment]
@jctrouble Thanks a lot for this great bug description - it was really very simple to reproduce it!

As easy as it was to reproduce, fixing took quite some time, but here we go: Do you mind installing the `cycle-conditions` branch and test if that improves your own project?

**Comment by jctrouble** [comment]
Unfortunately, when I went to train, two things happened:

1. The script hung for a _long_ time (O(1-2min)) at `structures.py:222` (the copying of the story steps)
1. The stack [blew out](https://gist.github.com/jctrouble/ec62d819b5da78761142520b04179b6a) doing the topo sort

**Comment by tmbo** [comment]
Ok I gave it another try. Do you mind trying the latest version from that branch @jctrouble ?

**Comment by jctrouble** [comment]
Sorry for the slow response, this change appears to fix the issue on my end! Thanks for all the effort!

The only change I needed to make was in `generator.py`. This line:

`for k, step in story_steps.items()[:]:`

doesn't play well with python3, but this works:

`for k, step in list(story_steps.items()):`

**Comment by jctrouble** [comment]
Additionally, I see number of warnings like this:

`WARNING:rasa_core.training.generator:Unsatisfied start checkpoint 'CYCLE_CONNECT_0d71345aa06644119f327df1332e86e6' in block 'Foo'`

It's unclear to me if it's indicative of a problem or not.
